crates/typst/src/math/lr.rs

```diff
@@ -64,7 +64,25 @@ impl LayoutMath for Packed<LrElem> {
             .relative_to(2.0 * max_extent);
 
         // Scale up fragments at both ends.
-        match fragments.as_mut_slice() {
+        // We calculate the start and end indices; I tried using slices and
+        // `split_{first,last}_mut` instead, but that ran into lifetime issues,
+        // and we need the indices anyway for the `fragments.retain` call.
+        let mut start_idx = 0;
+        let mut end_idx = fragments.len();
+        while start_idx < fragments.len() {
+            if !fragments[start_idx].is_ignorant() {
+                break;
+            }
+            start_idx += 1;
+        }
+        while end_idx > start_idx {
+            if !fragments[end_idx - 1].is_ignorant() {
+                break;
+            }
+            end_idx -= 1;
+        }
```

Comment:
I think this would be better with the iterator style:
```suggestion
        let start_idx = fragments
            .iter()
            .position(|f| !f.is_ignorant())
            .unwrap_or(fragments.len());
        let end_idx = fragments
            .iter()
            .skip(start_idx)
            .rposition(|f| !f.is_ignorant())
            .map_or(start_idx, |i| start_idx + i + 1);
```

Reply:
Better in what respect? It’s more work for the optimizer while being no less error-prone than a while-loop.

Reply:
Sorry, "better" is arguably not a good word here, it's more just personal preference and that (I think) the iterator pattern is usually preferred over loops throughout the code base. This is of course subjective, and performance aspects are probably negligible anyway.

Reply:
It's just a matter of preference, but I prefer the iterator style as well.

Reply:
Mutable code is _always_ more error-prone than immutable code. One can argue whether it is easier for newcomers or not -- which I also disagree, btw -- but, if there's no performance difference, it's quite hard to argue against immutability.

Reply:
> Mutable code is _always_ more error-prone than immutable code.

\[citation needed\]

In this case, the iterator-based code has an extra `+ start_idx` to compensate for the skip, plus an extra `+ 1` to avoid a fencepost error.

In any case, changing the loops to use the iterator-based approach adds 64 bytes to the executable on release mode (33300968 vs. 33300904 on the `x86_64-unknown-linux-gnu` target). I also tried comparing the [assembly outputs](https://gist.github.com/bluebear94/95c7e4d5635f247ebb8087134ec713e8) of the two, and it seems that some bound checks aren’t being elided in the iterator approach. But to be sure of the performance impacts, we would need to use something like [iai-callgrind](https://github.com/iai-callgrind/iai-callgrind) to eke out small differences like these.

Reply:
The iterator approach turns out to be more optimizable for `start_idx` ([test code for posterity](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=ad2f10b8a2aa1cc6d40cad8f1c411d05)).

Reply:
The code has become really complex now. It is far off the hot path, so I think the discussion about performance is a bit misguided (adding a comment about how it optimizes actually feels counter-productive to me as it brings focus to performance even though it doesn't really matter here.). 

I would like to just pick the simplest/clearest. My proposed solution would be:
```rs
let start_skip = fragments
    .iter()
    .filter(MathFragment::is_ignorant)
    .count();

let end_skip = fragments
    .iter()
    .rev()
    .filter(MathFragment::is_ignorant)
    .count();

let inner = &fragments[start_skip..fragments.len() - end_skip];
```

Reply:
Not being on the “hot path” is not a sufficient reason to purposefully make a piece of code perform worse.

Reply:
It is when readability suffers. I'm not sure what else to say here anymore.

Reply:
IMO my code isn’t a big sacrifice in readability for the performance gain, especially when it might be called many times when compiling a math-heavy document (given that anything in some sort of brackets is implicitly wrapped in an `LrElem`).